llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -1589,80 +1610,121 @@ bool SimplifyCFGOpt::hoistCommonCodeFromSuccessors(BasicBlock *BB,
   }
 
   bool Changed = false;
+  auto *SuccIterPairBegin = SuccIterPairs.begin();
+  SuccIterPairBegin++;
+  auto OtherSuccIterPairRange =
+      iterator_range(SuccIterPairBegin, SuccIterPairs.end());
+  auto OtherSuccIterRange = make_first_range(OtherSuccIterPairRange);
+  using InstrFlagPair = std::pair<Instruction *, unsigned>;
+  SmallVector<DenseMap<llvm::hash_code, InstrFlagPair>, 2> OtherSuccessorsHash;
+
+  for (auto BBItrPair : OtherSuccIterRange) {
+    // Fill the hashmap for every other successor
+    DenseMap<llvm::hash_code, InstrFlagPair> hashMap;
+    unsigned skipFlag = 0;
+    Instruction *I = nullptr;
+    do {
+      I = &*BBItrPair;
+      skipFlag |= skippedInstrFlags(I);
```

Comment:
Does `SkipFlag` need to be updated in each loop?

Reply:
Actually yes, because I keep track two SkipFlags for each successor, one created for every instruction when I fill the hashmap, the other one created for that successor and gets updated while I'm iterating through first successor instructions (same as previous approach). The second one will be valid as long as I have not hoisted two instructions that are not in the same level.

Reply:
> Actually yes, because I keep track two SkipFlags for each successor, one created for every instruction when I fill the hashmap, the other one created for that successor and gets updated while I'm iterating through first successor instructions (same as previous approach). The second one will be valid as long as I have not hoisted two instructions that are not in the same level.

When hoisting a set of instructions with different levels, should we update it?

Reply:
No we shouldn't. That's why I check `SameLevelHoist` before updating skipflags.

Reply:
When hoisting an instruction of a different level, should we update the skip flag of the subsequent instructions?

Reply:
Right, so actually yes. When we hoist an instruction, we might introduce new opportunities for subsequent instructions (they might become legal to hoist while they weren't before). I currently do not update flags because of the compile time overhead it brings. If you think it should be done, I will be happy to do that.

Reply:
I guess the compile time overhead isn't there. But I think you can submit a separate PR after that. We don't require a PR to accomplish everything. :)